src/AbstractGenerator.h

```diff
@@ -0,0 +1,194 @@
+     * build_pipeline() is called)
+     *
+     * This can be called multiple times, but only prior to build_pipeline().
+     *
+     * CALL-AFTER: nona
+     * CALL-BEFORE: build_pipeline
+     */
+    // @{
+    virtual void set_generatorparam_value(const std::string &name, const std::string &value) = 0;
+    virtual void set_generatorparam_value(const std::string &name, const LoopLevel &loop_level) = 0;
+    // @}
+
+    /** Build and return the Pipeline for this AbstractGenerator. This method should be called
+     * only once per instance.
+     *
+     * CALL-AFTER: set_generatorparam_value, bind_input
+     * CALL-BEFORE: get_parameters_for_input, get_funcs_for_output, get_external_code_map
+     */
+    virtual Pipeline build_pipeline() = 0;
+
+    /** Given the name of an input, return the Parameter(s) for that input.
+     * (Most inputs will have exactly one, but inputs that are declared as arrays
+     * will have multiple.)
+     *
+     * CALL-AFTER: build_pipeline
+     * CALL-BEFORE: none
+     */
+    virtual std::vector<Parameter> get_parameters_for_input(const std::string &name) = 0;
+
+    /** Given the name of an output, return the Func(s) for that output.
+     * (Most outputs will have exactly one, but outputs that are declared as arrays, or that return Tuples,
+     * will have multiple.)
+     *
+     * Must be called after build_pipeline(), since the output Funcs will be undefined prior to that.
+     *
+     * CALL-AFTER: build_pipeline()
+     * CALL-BEFORE: none
+     */
+    virtual std::vector<Func> get_funcs_for_output(const std::string &name) = 0;
+
+    /** Return the ExternsMap for the Generator, if any.
+     *
+     * CALL-AFTER: build_pipeline()
+     * CALL-BEFORE: n/a
+     */
+    virtual ExternsMap get_external_code_map() = 0;
+
+    /** Rebind all the inputs in a single call. The vector of StubInputs is assumed to have
+     * the same size and ordering as the list of inputs. Basic type-checking is done to ensure
+     * that inputs are still sane (e.g. types, dimensions, etc must match expectations).
+     *
+     * This call is infrequently used, and is only useful in JIT or Stub situations;
+     * this allows you to bind the inputs to be Funcs, Buffers, etc from another Halide code
+     * fragment, allowing inlining of multiple fragments together.
+     *
+     * CALL-AFTER: n/a
+     * CALL-BEFORE: build_pipeline
+     */
+    // @{
+    virtual void bind_input(const std::string &name, const std::vector<Parameter> &v) = 0;
+    virtual void bind_input(const std::string &name, const std::vector<Func> &v) = 0;
+    virtual void bind_input(const std::string &name, const std::vector<Expr> &v) = 0;
+    // @}
+
+    /** Emit a Generator Stub (.stub.h) file to the given path. Not all Generators support this.
+     *
+     * If you call this method, you should not call any other AbstractGenerator methods
+     * on this instance, before or after this call.
```

Comment:
Thank you @steven-johnson for providing the `AbstractGenerator` base class. I see it as an opportunity to port any custom AOT implementations to the Generator interface. Here's one example to emit FPGA accelerator code for HW synthesis, which is incompatible to the `Generator<>` interface:

https://github.com/jingpu/Halide-HLS/blob/HLS/apps/hls_examples/conv_hls/pipeline.cpp#L131-L135

Quick question: if I were to implement `emit_cpp_stub` function
```c++
class MyGenerator final : public AbstractGenerator {
...
bool MyGenerator::emit_cpp_stub(const string& path) final {

        output.compile_to_lowered_stmt(path + "pipeline_host.ir.html", args, HTML);
        output.compile_to_header(path + "pipeline_host.h", args, "pipeline_native");
        output.compile_to_object(path + "pipeline_host.o", args, "pipeline_native");

Target hls_target = get_target_from_environment();
        hls_target.set_feature(Target::CPlusPlusMangling);
        output.compile_to_lowered_stmt(path + "pipeline_target.ir.html", args, HTML, hls_target);
        output.compile_to_hls(path + "pipeline_target.cpp", args, "pipeline_hls", hls_target);
        output.compile_to_header(path + "pipeline_target.h", args, "pipeline_hls", hls_target);
}
};
```

Can I call define the following Makefile target to emit the AOT binaries? How about the cpp stub?
```Make
pipeline_target.cpp: pipeine.generator
    ./$< -g pipeline_name target=zynq -e 'html,h,cpp' -o . && echo "$@ is generated"

pipeline_host.o: pipeline.generator
    ./$< -g pipeline_name target=host -e 'html,h,o' -o . && echo "$@ is generated"
```

Reply:
> Thank you @steven-johnson

Well, it hasn't landed yet, but I hope it will soon :-)

> which is incompatible to the `Generator<>` interface

Interesting, can you share a bit more about how and why your work is incompatible? (It's always good to learn about how and why and existing design isn't adequate.)

> if I were to implement emit_cpp_stub function

The implementation you have there isn't actually emitting a cpp_stub -- a cpp_stub is a very weird chunk of code that is intended to make a Generator more usable in JIT mode, or when composing Generators together. (It's very rarely used and frankly should be removed if downstream users ever stop using it.) What are you trying to accomplish here?

BTW #2: I *highly* recommend you *don't* examine any environment variables in your Generator; that includes `get_target_from_environment()`; you can see the Target that was specified for your generator simply by calling its `get_target()` method.

Reply:
> Interesting, can you share a bit more about how and why your work is incompatible? (It's always good to learn about how and why and existing design isn't adequate.)

Sorry, I take it back. I realized the original author of the forked project implemented `Pipeline::compile_to_hls`, but never got that far to integrate it into `Generator<>`. So, if I were to schedule an image processing pipeline with `Generator<>::schedule()`, I can schedule `target=host` but not `target=zynq`.

https://github.com/jingpu/Halide-HLS/blob/905d2f2ad560246673ba3a84b8a6d8be308e481f/src/Pipeline.h#L148-L163

> not emitting a cpp_stub.

Got it. Thanks for the reminder.

> don't examine any environment variables.

My bad. when I drafted the pseudo code, I copied it out of the JIT code from the original author. As I mentioned, it would be ideal if I can organize the JIT code (link below) into either `Generator<>` or `AbstractGenerator`. I simply don't know how it is done, starting from the developer provided `Pipeline::compile_to_*` API in `src/pipeline.h`.

https://github.com/jingpu/Halide-HLS/blob/HLS/apps/hls_examples/conv_hls/pipeline.cpp
https://github.com/jingpu/Halide-HLS/blob/HLS/apps/hls_examples/unsharp_hls/pipeline.cpp#L113-L153

I can create a new issue for a more in depth discussion. Again, I appreciate your work on `AbstractGenerator` class.

Reply:
> I can create a new issue for a more in depth discussion.

Yes, please do.